DAQ/src/StrawHitRecoFromFragments_module.cc

```diff
@@ -0,0 +1,360 @@
+  //auto sdH = event.getValidHandle(_sdctoken);
+  //const StrawDigiCollection& sdcol(*sdH);
+
+  //const StrawDigiADCWaveformCollection *sdadccol(0);
+  //if (_fittype != TrkHitReco::FitType::firmwarepmp) {
+  //  auto sdawH = event.getValidHandle(_sdadctoken);
+  //  sdadccol = sdawH.product();
+  //}
+
+  const mu2e::CaloClusterCollection* caloClusters(0);
+  if(_usecc){
+    auto ccH = event.getValidHandle(_ccctoken);
+    caloClusters = ccH.product();
+  }
+
+  art::Handle<mu2e::ProtonBunchTime> pbtHandle;
+  auto pbtH = event.getValidHandle(_pbttoken);
+  const mu2e::ProtonBunchTime& pbt(*pbtH);
+  pbtOffset = pbt.pbtime_;
+
+  mu2e::TrackerStatus const& trackerStatus = _trackerStatus_h.get(event.id());
+  std::unique_ptr<mu2e::StrawHitCollection> shCol;
+  if(_writesh){
+    shCol = std::unique_ptr<mu2e::StrawHitCollection>(new mu2e::StrawHitCollection);
+  }
+  std::unique_ptr<mu2e::ComboHitCollection> chCol(new mu2e::ComboHitCollection());
+
+  if (!trkFragments.isValid()) {
+    std::cout << "[StrawAndCaloDigisFromFragments::produce] found no Tracker fragments!"
+      << std::endl;
+    if(_writesh)event.put(std::move(shCol));
+    event.put(std::move(chCol));
+    return;
+  }
+  size_t numTrkFrags = trkFragments->size();
+
+  if(_writesh){
+    shCol->reserve(numTrkFrags);
+  }
+  chCol->reserve(numTrkFrags);
+
+  mu2e::StrawHitRecoUtils shrUtils(pbtOffset, _fittype, _npre, _invnpre, _invgainAvg, _invgain,
+    _diagLevel, _maxiter, _mask, nplanes, npanels, _writesh, _minT, _maxT, _minE, _maxE, _filter, _flagXT,
+    _ctE, _ctMinT, _ctMaxT, _usecc, _clusterDt, numTrkFrags);
+
+
+  for (size_t idx = 0; idx < numTrkFrags; ++idx) {
+    analyze_tracker_((*trkFragments)[idx], shCol, chCol, shrUtils, trackerStatus, srep, caloClusters, tt);
+  }
+
+  if(_writesh)event.put(std::move(shCol));
+  event.put(std::move(chCol));
+}
+
+
+void art::StrawHitRecoFromFragments::analyze_tracker_(
+    const artdaq::Fragment& f, std::unique_ptr<mu2e::StrawHitCollection> const& shCol,
+    std::unique_ptr<mu2e::ComboHitCollection> const& chCol, mu2e::StrawHitRecoUtils &shrUtils, 
+    mu2e::TrackerStatus const& trackerStatus, mu2e::StrawResponse const& srep,
+    const mu2e::CaloClusterCollection* caloClusters, mu2e::Tracker const& tt
+    ) {
+
+  mu2e::TrackerFragment cc(f);
+
+  if (_diagLevel > 1) {
+    std::cout << std::endl;
+    std::cout << "TrackerFragment: ";
+    std::cout << "\tBlock Count: " << std::dec << cc.block_count() << std::endl;
+    std::cout << "\tByte Count: " << f.dataSizeBytes() << std::endl;
+    std::cout << std::endl;
+    std::cout << "\t"
+      << "====== Example Block Sizes ======" << std::endl;
+    for (size_t i = 0; i < 10; i++) {
+      if (i < cc.block_count()) {
+	std::cout << "\t" << i << "\t" << cc.blockSizeBytes(i) << std::endl;
+      }
+    }
+    std::cout << "\t"
+      << "=========================" << std::endl;
+  }
+
+  for (size_t curBlockIdx = 0; curBlockIdx < cc.block_count(); curBlockIdx++) {
```

Comment:
This looks like a packet diagnostic analyzer function.  Should we have a separate analyzer module just for diagnostics?  Or is there a reason it's inside this producer.  I'm also not clear on why this block is double-protected (compile flag and runtime flag).

Reply:
This is just copied from the corresponding DAQ code, I'll take it out